### ISSUE TITLE
NATS microservice and client not reconnecting

### ISSUE THREAD

**Issue by MickL** [nestjs/nest#6684]
## Current behavior
If a microservice app or client can't connect to NATS server, or disconnects after connected, it throws an error but never tries to reconnect.

## Input Code
```ts
const app = await NestFactory.createMicroservice<MicroserviceOptions>(AppModule, {
    transport: Transport.NATS,
    options:   {
      url:   'nats://localhost:4222',
      maxReconnectAttempts: -1,
    },
  });
```

## Expected behavior
It should reconnect according to the [connect options](https://github.com/nats-io/nats.js/tree/v1.4.12#connect-options) which sets `maxReconnectAttempts` and `reconnectTimeWait`.

## Further things
- The app never starts if there is no connection, instead it should retry until connected and then continue to start
- If connection is lost after initially connected, `ClientProxy.send()` throws an error on first call, but then does not throw an error anymore. It shows an log in console "NatsError: Connection closed" but does not throw an error.
- I dont see any hooks for logging. E.g. when the connection is lost I want to log an error and when the connection is back I want to log that, too. -> EDIT: Hooks are there but not accessible, see #6692

## Environment

<pre><code>
Nest version: 8.0.0-alpha.3
 
For Tooling issues:
- Node version: 15.6.0
- Platform:  macOS Big Sur

**Comment by Tony133** [comment]
Hi @MickL, I recommend that you leave a minimal reproduction of a clonable git repository

**Comment by kamilmysliwiec** [comment]
Please provide a minimum reproduction repository.

**Comment by MickL** [comment]
I created two apps: One for the client and one for a microservice. They are not meant to work together, just to test each separately: https://github.com/MickL/nestjs-nats

For the client I need to correct myself: It only happens when using `.send()` not `.emit()`. 

For `.emit()` using `maxReconnectAttempts: -1` it never throws an error but is actually waiting for a reconnect. So everything works as expected,

**Reproduction client `.send()`:**
1. While Docker is running, run `yarn docker-start`
2. Run `yarn start client`
3. `curl localhost:3000` -> 200
4. Stop NATS: `yarn docker-stop`
5. `curl localhost:3000` -> 500, exception is thrown
6. `curl localhost:3000` -> 200 but should return 500, no exception is thrown in `.send()`
7. Stop NATS: `yarn docker-start`
8. `curl localhost:3000` -> 200 but client does not reconnect, no exception is thrown

For "5." it is either okay like this OR it should have the same behavior as .emit() that just waits for reconnect.

**Reproduction microservice**
1. Dont have NATS running, `yarn docker-stop`
2. Run `yarn start microservice`
3. App does not start: "NatsError: Could not connect to server: Error: connect ECONNREFUSED 127.0.0.1:4222"
4. Start NATS, `yarn docker-start`
5. Nothing happens, app does not connect and never starts

**Comment by ceoro9** [comment]
@MickL , I tried to follow the steps you provided and here is the results I got:

* Running nats, client and microservice - everything works smoothly. Here is the nats connection params after the requests:
```
  reconnects: 0,
  connected: true,
  wasConnected: true,
  reconnecting: false,
  pending: [],
  pout: 0,
```

* Then I have stopped nats and made `curl localhost:3000` request a couple of times. And every time 500 is returned, because of rxjs timeout operator. `curl localhost:3000/emit` works exactly the same.

`curl localhost:3000`
```
// 1st request
reconnects: 4,
connected: false,
wasConnected: true,
reconnecting: true,
pending: [],
pout: 0,

// 2nd request
reconnects: 9,
connected: false,
wasConnected: true,
reconnecting: true,
pending: [
  'PUB something _INBOX.X8AHDFQF0BV9OTCDP01ZUZ.X8AHDFQF0BV9OTCDP020BN 81\r\n' +
    '{"pattern":"something","data":"data","id":"db0fe986-5e7c-4c4e-9477-d0b5885da93c"}\r\n'
],
```

`curl localhost:3000/emit`
```
reconnects: 16,
connected: false,
wasConnected: true,
reconnecting: true,
pending: [
  'PUB something 33\r\n{"pattern":"something","data":{}}\r\n',
  'PING\r\n'
],
```

As you see from connection logs, nats driver actually keeps trying to reconnect as it is supposed to do.

So to summarize, my answer to the first part of your message is **works on my machine** (Manjaro Linux x86_64, 5.10.2-2-MANJARO).

---

Regarding the second part of your message. I think nats driver reconnecting mechanism works only if connection has even been setup before in the current execution(this is just a guess, based on `wasConnected` connection param).

If first attempt has failed, then driver's work is literally over. So that's why microservice execution fails right after its launch, because it tries to connect to nats immediately, unlike client, which does it when the first request comes.

**Comment by MickL** [comment]
Its funny because for .send at first I had a hard time reproducing again, too. But now I got it and definitely there seems to be a bug or something. It works as expected if there initially was a connection but it does not work if there was initially no connection. In that case the first `.send()` throws an error, as expected, but then it is returning a string "NatsError: Connection closed" and thats why no error is thrown.

I updated my repo to display this and I updated the instructions above.

**Comment by MickL** [comment]
@ceoro9 

>  you have to provide waitOnFirstConnect param to connection options(as comment says in source code below). Then driver will work as you expected.

If thats the case then there is still the bug: The first .send() correctly throws an error but all other .send() dont throw an error anymore and instead return an error string as the response.

**Comment by ceoro9** [comment]
@MickL 

> I checked and yes this is actually what I expected: The microservice app is not crashing on startup, instead it will wait for the connection. Also the client is working as expected now! I added this to the docs, thank you very much!
👍

>  the first .send() correctly throws an error but all other .send() dont throw an error anymore and instead return an error string as the response. This can be tested with my reproduction steps from above.
I were able to reproduce the issue, preparing a fix.

**Comment by MickL** [comment]
As Nest.js is using nats.js 2.0 it may be interesting to see if it this is the same over there.

**Comment by ceoro9** [comment]
hey, @kamilmysliwiec, can you please review my PR?

Soon I will update you with the status of that issue in `nats.js` 2.0

**Comment by ceoro9** [comment]
version `8.0.0-alpha.1` does not have such issue, because nats error is handled explicitly.
```typescript
 public createSubscriptionHandler(
    packet: ReadPacket & PacketId,
    callback: (packet: WritePacket) => any,
  ) {
    return (error: unknown | undefined, natsMsg: NatsMsg) => {
      // HERE
      if (error) {
        return callback({
          err: error,
        });
      }
      const rawPacket = natsMsg.data;
      const message = this.deserializer.deserialize(rawPacket);
      if (message.id && message.id !== packet.id) {
        return undefined;
      }
      const { err, response, isDisposed } = message;
      if (isDisposed || err) {
        return callback({
          err,
          response,
          isDisposed: true,
        });
      }
      callback({
        err,
        response,
      });
    };
  }
```

**Comment by MickL** [comment]
Thanks for all your afforts!

**Comment by kamilmysliwiec** [comment]
Thanks @ceoro9! Let's track this here https://github.com/nestjs/nest/pull/6696

**Comment by MickL** [comment]
Personally what I dont like is that neither Microservices nor ProxyClients log any connection errors. In comparison `@nestjs/mongoose` does this very nicely:

<img width="592" alt="Bildschirmfoto 2021-03-26 um 11 59 50" src="https://user-images.githubusercontent.com/5103210/112622198-00050900-8e2b-11eb-9e57-d2a01abeac4d.png">

And as the natsClient is not public I cant hook into the events and log myself.
